File: python/mirage/core/notion/resolve.py
from typing import Any

from mirage.accessor.notion import NotionAccessor
from mirage.core.hierarchy.scope import ScopeMatch
from mirage.core.notion.client import NotionAPIError
from mirage.core.notion.pages import get_page
from mirage.core.notion.pathing import page_dirname
from mirage.utils.errors import enoent


async def resolve_row(accessor: NotionAccessor, match: ScopeMatch,
                      virtual: str) -> dict[str, Any]:
    """Fetch and validate the row named by a path, including descendants.

    Args:
        accessor (NotionAccessor): mount client.
        match (ScopeMatch): classified path with a distinct row identity.
        virtual (str): path to name in a refusal.
    """
    try:
        page = await get_page(accessor.config,
                              match.slots["row_id"],
                              session=accessor.pool)
    except NotionAPIError as exc:
        if exc.status == 404 or exc.code == "validation_error":
            raise enoent(virtual) from exc
        raise
    parent = page.get("parent", {})
    name = f'{match.slots["row"]}__{match.slots["row_id"]}'
    if (parent.get("data_source_id") != match.slots["data_source_id"]
            or page.get("in_trash") or page.get("archived")
            or page_dirname(page) != name):
        raise enoent(virtual)
    return page


async def guard_row(accessor: NotionAccessor, match: ScopeMatch,
                    virtual: str) -> None:
    """Validate a containing row when this path has one.

    Args:
        accessor (NotionAccessor): mount client.
        match (ScopeMatch): classified page or row path.
        virtual (str): path to name in a refusal.
    """
    if "row_id" in match.slots:
        await resolve_row(accessor, match, virtual)

File: typescript/packages/core/src/core/notion/resolve.ts
import type { NotionAccessor } from '../../accessor/notion.ts'
import { enoent } from '../../utils/errors.ts'
import type { ScopeMatch } from '../hierarchy/scope.ts'
import { NotionAPIError } from './client.ts'
import { pageSegmentName } from './normalize.ts'
import { getPage } from './pages.ts'

/** Fetch and validate the row named by a path, including descendants. */
export async function resolveRow(
  accessor: NotionAccessor,
  match: ScopeMatch,
  virtual: string,
): Promise<Record<string, unknown>> {
  let page: Record<string, unknown>
  try {
    page = await getPage(accessor.transport, match.slots.row_id ?? '')
  } catch (err) {
    if (err instanceof NotionAPIError && (err.status === 404 || err.code === 'validation_error')) {
      throw enoent(virtual)
    }
    throw err
  }
  const parent = page.parent as Record<string, unknown> | undefined
  const name = `${match.slots.row ?? ''}__${match.slots.row_id ?? ''}`
  if (
    parent?.data_source_id !== match.slots.data_source_id ||
    page.in_trash === true ||
    page.archived === true ||
    pageSegmentName(page) !== name
  )
    throw enoent(virtual)
  return page
}

/** Validate a containing row when this path has one. */
export async function guardRow(
  accessor: NotionAccessor,
  match: ScopeMatch,
  virtual: string,
): Promise<void> {
  if (match.slots.row_id !== undefined) await resolveRow(accessor, match, virtual)
}
